prometheus-connector/src/main/java/org/apache/flink/connector/prometheus/sink/PrometheusSinkWriter.java

```diff
@@ -0,0 +1,267 @@
+        private final SinkCounters counters;
+
+        public ResponseCallback(
+                int timeSeriesCount,
+                long sampleCount,
+                SinkCounters counters,
+                Consumer<List<Types.TimeSeries>> requestResult) {
+            this.timeSeriesCount = timeSeriesCount;
+            this.sampleCount = sampleCount;
+            this.requestResult = requestResult;
+            this.counters = counters;
+        }
+
+        @Override
+        public void completed(SimpleHttpResponse response) {
+            if (RemoteWriteResponseClassifier.isSuccessResponse(response)) {
+                LOG.debug(
+                        "{},{} - successfully posted {} time-series, containing {} samples",
+                        response.getCode(),
+                        response.getReasonPhrase(),
+                        timeSeriesCount,
+                        sampleCount);
+                counters.inc(NUM_SAMPLES_OUT, sampleCount);
+                counters.inc(NUM_WRITE_REQUESTS_OUT);
+            } else {
+                counters.inc(NUM_SAMPLES_DROPPED, sampleCount);
+                counters.inc(NUM_WRITE_REQUESTS_PERMANENTLY_FAILED);
+
+                String responseBody = response.getBodyText();
+                if (RemoteWriteResponseClassifier.isNonRetriableErrorResponse(response)) {
+                    LOG.warn(
+                            "{},{} {} (discarded {} time-series, containing {} samples)",
+                            response.getCode(),
+                            response.getReasonPhrase(),
+                            responseBody,
+                            timeSeriesCount,
+                            sampleCount);
+                    counters.inc(NUM_SAMPLES_NON_RETRIABLE_DROPPED, sampleCount);
+                } else if (RemoteWriteResponseClassifier.isRetriableErrorResponse(response)) {
+                    LOG.warn(
+                            "{},{} {} (after retry limit reached, discarded {} time-series, containing {} samples)",
+                            response.getCode(),
+                            response.getReasonPhrase(),
+                            responseBody,
+                            timeSeriesCount,
+                            sampleCount);
+                    counters.inc(NUM_SAMPLES_RETRY_LIMIT_DROPPED, sampleCount);
+                }
+            }
+
+            // Never re-queue requests
+            requestResult.accept(Collections.emptyList());
+        }
+
+        @Override
+        public void failed(Exception ex) {
+            LOG.warn(
+                    "Exception executing the remote-write (discarded {} time-series containing {} samples)",
+                    timeSeriesCount,
+                    sampleCount,
+                    ex);
+            counters.inc(NUM_SAMPLES_DROPPED, sampleCount);
+            counters.inc(NUM_WRITE_REQUESTS_PERMANENTLY_FAILED);
+        }
```

Comment:
Should we retry these failures / Propagate the exception up to transition the job to failed?

Reply:
Changed the general behaviour (see comment below)